File: backend/src/models/Order.js
const mongoose = require('mongoose');

const orderItemSchema = new mongoose.Schema({
  menuItem: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'MenuItem',
    required: true,
  },
  name: String,
  quantity: {
    type: Number,
    required: true,
    min: 1,
  },
  price: Number, // unit price in BDT paisa
  customizations: [{
    name: String,
    option: String,
    price: Number,
  }],
  specialInstructions: String,
});

const orderSchema = new mongoose.Schema({
  orderNumber: {
    type: String,
    unique: true,
  },
  user: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'User',
    required: true,
  },
  restaurant: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'Restaurant',
    required: true,
  },
  rider: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'Rider',
  },
  items: [orderItemSchema],
  deliveryAddress: {
    label: String,
    houseNo: String,
    roadNo: String,
    area: String,
    thana: String,
    district: String,
    postalCode: String,
    landmark: String,
    location: {
      type: {
        type: String,
        default: 'Point',
      },
      coordinates: [Number],
    },
  },
  subtotal: {
    type: Number,
    required: true, // in BDT paisa
  },
  deliveryFee: {
    type: Number,
    default: 0,
  },
  vat: {
    type: Number,
    default: 0, // 5% in Bangladesh
  },
  discount: {
    type: Number,
    default: 0,
  },
  total: {
    type: Number,
    required: true,
  },
  paymentMethod: {
    type: String,
    enum: ['bkash', 'nagad', 'rocket', 'card', 'cod'],
    required: true,
  },
  paymentStatus: {
    type: String,
    enum: ['pending', 'paid', 'failed', 'refunded'],
    default: 'pending',
  },
  paymentRef: String, // Payment gateway reference
  status: {
    type: String,
    enum: ['pending', 'confirmed', 'preparing', 'ready', 'picked_up', 'on_the_way', 'delivered', 'cancelled'],
    default: 'pending',
  },
  statusHistory: [{
    status: String,
    timestamp: {
      type: Date,
      default: Date.now,
    },
    note: String,
  }],
  specialInstructions: String,
  scheduledFor: Date,
  estimatedDelivery: Date,
  actualDelivery: Date,
  cancellationReason: String,
  rating: {
    foodRating: Number,
    deliveryRating: Number,
    review: String,
  },
}, {
  timestamps: true,
});

// Generate order number before saving
orderSchema.pre('save', async function(next) {
  if (!this.orderNumber) {
    const count = await mongoose.model('Order').countDocuments();
    this.orderNumber = `KL${Date.now().toString().slice(-8)}${String(count + 1).padStart(4, '0')}`;
  }
  next();
});

orderSchema.index({ user: 1, createdAt: -1 });
orderSchema.index({ restaurant: 1, status: 1 });
orderSchema.index({ rider: 1, status: 1 });

module.exports = mongoose.model('Order', orderSchema);

File: restaurant-app/src/main/java/com/noor/khabarlagbe/restaurant/domain/model/Order.kt
package com.noor.khabarlagbe.restaurant.domain.model

data class Order(
    val id: String,
    val orderNumber: String,
    val customerId: String,
    val customerName: String,
    val customerPhone: String,
    val customerAddress: Address,
    val items: List<OrderItem>,
    val subtotal: Double,
    val deliveryFee: Double,
    val packagingCharge: Double,
    val discount: Double,
    val total: Double,
    val status: OrderStatusEnum,
    val paymentMethod: String,
    val paymentStatus: String,
    val specialInstructions: String?,
    val estimatedPrepTime: Int,
    val riderId: String?,
    val riderName: String?,
    val riderPhone: String?,
    val timeline: List<OrderTimeline>,
    val createdAt: String,
    val updatedAt: String
)

data class OrderItem(
    val id: String,
    val menuItemId: String,
    val name: String,
    val nameEn: String?,
    val quantity: Int,
    val price: Double,
    val totalPrice: Double,
    val customizations: List<OrderCustomization>?,
    val specialInstructions: String?
)

data class OrderCustomization(
    val name: String,
    val option: String,
    val price: Double
)

data class OrderTimeline(
    val status: String,
    val timestamp: String,
    val note: String?
)

enum class OrderStatusEnum {
    PENDING,
    ACCEPTED,
    PREPARING,
    READY,
    PICKED_UP,
    DELIVERED,
    CANCELLED,
    REJECTED;
    
    companion object {
        fun fromString(status: String): OrderStatusEnum {
            return when (status.lowercase()) {
                "pending" -> PENDING
                "accepted" -> ACCEPTED
                "preparing" -> PREPARING
                "ready" -> READY
                "picked_up" -> PICKED_UP
                "delivered" -> DELIVERED
                "cancelled" -> CANCELLED
                "rejected" -> REJECTED
                else -> PENDING
            }
        }
    }
    
    fun toApiString(): String {
        return when (this) {
            PENDING -> "pending"
            ACCEPTED -> "accepted"
            PREPARING -> "preparing"
            READY -> "ready"
            PICKED_UP -> "picked_up"
            DELIVERED -> "delivered"
            CANCELLED -> "cancelled"
            REJECTED -> "rejected"
        }
    }
    
    fun displayNameBn(): String {
        return when (this) {
            PENDING -> "অপেক্ষমান"
            ACCEPTED -> "গৃহীত"
            PREPARING -> "প্রস্তুত হচ্ছে"
            READY -> "প্রস্তুত"
            PICKED_UP -> "পিকআপ হয়েছে"
            DELIVERED -> "ডেলিভারি সম্পন্ন"
            CANCELLED -> "বাতিল"
            REJECTED -> "প্রত্যাখ্যাত"
        }
    }
}
